### ISSUE TITLE
allowAccess rule validator

### ISSUE THREAD

**Issue by trombonekenny** [PrairieLearn/PrairieLearn#2622]
Add a validator (either run at sync or when viewing an assessment or course instance allowAccess page) to double-check access rules for correctness and that the sequence of the rules makes sense.

Things the validator might check:
- uids listed which are not part of the course (typo'd?)
- identify access rules further down the list that will never get triggered because the specific rules follow after generic ones
- rules that break other syntax (end_date before start_date)

**Comment by nwalters512** [comment]
Once #1640 lands, this would be great to add as a "warning".

**Comment by trombonekenny** [comment]
Bump for warning the instructor when certain access rules will never fire (specific rules lower than generic ones)

**Comment by reteps** [comment]
Is this validated at sync time via the UI yet?

**Comment by nwalters512** [comment]
What do you mean by "at sync time via the UI"?

**Comment by reteps** [comment]
Sorry, the "via the UI" qualifier there is not useful. I mean does the syncing code allow for those cases to occur without raising any errors.

**Comment by nwalters512** [comment]
The improper sequencing of dates is covered here:

https://github.com/PrairieLearn/PrairieLearn/blob/91d410eb85ea19cb7b434d70398624b01cd36c16/apps/prairielearn/src/sync/course-db.ts#L883-L887

Regarding the others:

- I'm actually not sure if we want to flag invalid UIDs during sync. That requires hitting the DB, and while that's certainly possible, it could slow down sync. We could potentially flag that in the UI while viewing the access rules for a particular thing? Still not sure on the value of even that though. (This would have to be disabled in dev mode since most UIDs won't exist there).
- For "access rules that will never get triggered", that seems like a sensible change and one that we can probably compute really quickly during sync. However, I can't immediately think of examples of this kind of thing. @trombonekenny, do you have any in mind that we can use in this discussion?

**Comment by mwest1066** [comment]
Overall I'm more in favor of #11236 than attempting to catch errors in the existing access rules. I feel like the problems we see in practice tend to be more complex (e.g., the recent CHEM exam that had a missing "active: false") and wouldn't be caught by simple checks on the current rules. Fundamentally, I think the problem is that our current access rules don't use the semantics that instructors use, making it hard for an instructor to understand whether they will get what they intend.

**Comment by reteps** [comment]
I'm in agreement -- the only reason I was looking at this PR was that I was seeing what issues had this mysterious "diff:medium" tag 😄

**Comment by trombonekenny** [comment]
I'm in favor of putting work into a different access system, but maybe this is still useful for catching what we would want to validate and alert the user? @jonatanschroeder, didn't you have an experimental assessment access page that tried to put a UX that interpreted things? Any thoughts that contribute here?

Two examples for access rules that never trigger are 1) specific rules (uids) after generic rules and 2) assessment access rules that don't work within the course instance access rule (and how/where to alert that). I guess today's Slack example of validating UID formatting (if not enrollment) would fit here too and that's why this is back in discussion?

The code to resolve this might be medium, but the instances where these things occur certainly aren't rare.

**Comment by nwalters512** [comment]
I just opened https://github.com/PrairieLearn/PrairieLearn/pull/11995 as a response to recent feedback on Slack, as this was straightforward to implement.

I think adding some kind of validation for course instance access dates would also be sensible. I'm less sure about other things like "specific rules after generic rules". We want a high signal-to-noise ratio, and I'm not sure how we can easily achieve that; there are plenty of valid ways of doing access rules that involve interleading UID rules with non-UID rules. Maybe spell out the kind of validation you'd like to do in greater detail?

**Comment by jonatanschroeder** [comment]
There is a preliminary evaluation of validation of rules that never get triggered in #3348, particularly in https://github.com/PrairieLearn/PrairieLearn/blob/a0ce40babfb404b16c8ed16d13d3755b543af5ea/apps/prairielearn/src/pages/instructorAssessmentAccess/instructorAssessmentAccess.js#L194-L200

In essence it builds a list of effective rules from the original access rules, and identifies any rules that don't apply anywhere. This is done by breaking down the rules into generic rules and rules that apply to a subset of students. Any time a rule has a range of dates that overlaps with a previous rule in the same set, it changes the effective start/end date according to the previous rule. For UID-specific rules, it applies them to all previous generic+specific rules, and any future generic rules apply both for generic and specific sets. If a resulting rule ends up with no application anywhere, it is flagged as having no application.

**Comment by nwalters512** [comment]
Modern access control has a stricter data model and features more built-in validations. Other class of issues (mistyped UIDs) are impossible in this new system.
